crema/qvalues.py

```diff
@@ -132,3 +136,213 @@ def _fdr2qvalue(scores, fdr):
+        LOGGER.debug("FALLBACK: pi0==1.0; all q-values will be 1.0")
+        fdrmod = np.full(num_targets, 1.0)  # all q-values are 1
+    elif pi0 < 0 or pi0 >= 1 or not np.isfinite(pi0):
+        raise ValueError(
+            f"Invalid pi0 estimate ({pi0}); unable to proceed FDR estimation!"
+        )
+    else:
+        fdrmod = calculate_mixmax_qval(
+            np.array(target_scores), np.array(decoy_scores), pi0
+        )
+
+    return (pi0, fdrmod)
+
+
+@nb.njit
+def estimate_pi0(pval_list):
+    """
+    Estimates pi0. Add description. TODO
+
+    Parameters
+    -------
+    pval_list : numpy.ndarray of float
+        A list of p-values sorted from smallest to largest.
+
+    Returns
+    -------
+    pi0 : float
+        Estimated pi_zero.
+    """
+    numLambda = 100
+    maxLambda = 0.5
+    numBoot = 100
+
+    # LOGGER.debug("pval_list=%s", pval_list)
+
+    n_pval = pval_list.size
+    lambda_list = []
+    pi0s_list = []
+    for idx in range(0, numLambda):
+        cur_lambda = ((idx + 1) / numLambda) * maxLambda
+
+        # Find the index of the first element in pValList
+        # that is > lambda.
+        start = np.searchsorted(pval_list, cur_lambda)
+        W1 = n_pval - start
+        pi0 = W1 / n_pval / (1.0 - cur_lambda)
+
+        if pi0 > 0.0:
+            lambda_list.append(cur_lambda)
+            pi0s_list.append(pi0)
+
+    # LOGGER.debug("%s", lambda_list)
+    # LOGGER.debug("%s", pi0s_list)
+
+    assert len(pi0s_list) != 0, (
+        "Error in the input data: "
+        "too good separation between target and decoy PSMs."
+    )
+
+    minPi0 = min(pi0s_list)
+
+    mse_list = np.zeros(len(pi0s_list))
+    max_size = 1000
+    # Examine which lambda level that is most stable under bootstrap
+    for i in range(0, numBoot):
+        # Create an array of bootstrapped p-values, and sort in ascending order.
+        num_draw = min(n_pval, max_size)
+        pBoot_list = np.random.choice(pval_list, size=num_draw, replace=True)
+        pBoot_list.sort()
+
+        for idx in range(0, len(lambda_list)):
+            start = np.searchsorted(pBoot_list, lambda_list[idx])
+            W1 = num_draw - start
+            pi0Boot = W1 / num_draw / (1.0 - lambda_list[idx])
+
+            # Estimated mean-squared error
+            mse_list[idx] += (pi0Boot - minPi0) * (pi0Boot - minPi0)
+
+    # Which index did the iterator go?
+    minIdx = np.argmin(mse_list)
+
+    # LOGGER.debug(f"Estimated pi0=%f at lambda=%f (MSE=%f)", pi0s_list[minIdx], lambda_list[minIdx], mse_list[minIdx])
+
+    pi0 = max(min(pi0s_list[minIdx], 1.0), 0.0)
+    return pi0
+
+
+@nb.njit
+def calculate_mixmax_qval(target_scores, decoy_scores, pi0):
+    """ """
+    # assert pi0 >= 0 and pi0 < 1
+
+    num_targets = target_scores.shape[0]
+    num_decoys = decoy_scores.shape[0]
+
+    h_w_le_z = np.zeros(num_decoys + 1)  # histogram for N_{w<=z}
+    h_z_le_z = np.zeros(num_decoys + 1)  # histogram for N_{z<=z}
```

Comment:
**nitpick (non-blocking):** I know they're probably from the paper, but these variable names are nonsensical to me.

Reply:
Correct, they are from the paper and those variables are the names found in Percolator. Can't decide whether to make them understandable or consistent with other code. Thoughts?

Reply:
Maybe add a comment describing what these variables are supposed to be?

Reply:
That might be for the best. I'd say at least just note where they came from.